subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -291,194 +290,302 @@ func (h *HivedAlgorithm) initReservations() {
+	pGpu *PhysicalCell,
+	vGpu *VirtualCell,
+	p CellPriority,
+	g *AlgoAffinityGroup) {
+
+	physicalPriority := p
+	if vGpu != nil {
+		preassignedNewlyBound := vGpu.GetPreAssignedCell().GetPhysicalCell() == nil
+		bindCell(pGpu, vGpu)
+		if preassignedNewlyBound {
+			// remove the allocated cell from the free list (possibly splitting cells)
+			h.removeCellFromFreeList(vGpu.GetPreAssignedCell().GetPhysicalCell())
+		}
+		vGpu.SetPriority(p)
+		updateUsedGpuNumAtPriority(vGpu, p, true)
 	} else {
-		return c.(*PhysicalCell)
+		physicalPriority = opportunisticPriority
+	}
+	pGpu.SetPriority(physicalPriority)
+	updateUsedGpuNumAtPriority(pGpu, physicalPriority, true)
+	pGpu.AddAffinityGroup(g)
+}
+
+// confirmReleasedGpu destroys the cell bindings, adds the physical cell back to the free list
+// (if necessary), and resets the priority.
+func (h *HivedAlgorithm) confirmReleasedGpu(pGpu *PhysicalCell, g *AlgoAffinityGroup) {
+	if vGpu := pGpu.GetVirtualCell(); vGpu != nil {
+		preassignedPhysical := vGpu.GetPreAssignedCell().GetPhysicalCell()
+		unbindCell(pGpu)
+		if vGpu.GetPreAssignedCell().GetPhysicalCell() == nil {
+			// add the released cell back to the free list (possibly merging cells)
+			h.addCellToFreeList(preassignedPhysical)
+		}
+		updateUsedGpuNumAtPriority(vGpu, vGpu.GetPriority(), false)
+		vGpu.SetPriority(freePriority)
+	}
+	updateUsedGpuNumAtPriority(pGpu, pGpu.GetPriority(), false)
+	pGpu.SetPriority(freePriority)
+	pGpu.DeleteAffinityGroup(g)
+}
+
+// removeCellFromFreeList removes a cell from the free cell list and splits its parent recursively if needed.
+func (h *HivedAlgorithm) removeCellFromFreeList(c *PhysicalCell) {
+	chain := c.GetChain()
+	terminate := false
+	for {
+		l := c.GetLevel()
+		parent := c.GetParent()
+		if parent != nil {
+			pp := parent.(*PhysicalCell)
+			if pp.IsSplit() {
+				terminate = true
+			} else {
+				h.freeCellList[chain][l] = append(h.freeCellList[chain][l], pp.GetChildren()...)
+				pp.SetSplit(true)
+			}
+		} else {
+			terminate = true
+		}
+		h.freeCellList[chain].remove(c, l)
+		if terminate {
+			break
+		} else {
+			c = parent.(*PhysicalCell)
+		}
 	}
 }
 
-func (h *HivedAlgorithm) getLevelByGpuNum(gpuNum int32, chain CellChain) CellLevel {
-	for l := CellLevel(1); l <= CellLevel(len(h.gpuNums[chain])); l++ {
-		if h.gpuNums[chain][l] >= gpuNum {
-			return l
+// addCellToFreeList adds a cell to the free cell list and merges its buddies recursively if needed.
+func (h *HivedAlgorithm) addCellToFreeList(c *PhysicalCell) {
+	chain := c.GetChain()
+	terminate := false
+	for {
+		l := c.GetLevel()
+		parent := c.GetParent()
+		if parent != nil {
+			allBuddyFree := true
+			for _, buddy := range parent.GetChildren() {
+				if buddy.(*PhysicalCell).GetVirtualCell() != nil {
+					allBuddyFree = false
+					break
+				}
+			}
+			if !allBuddyFree {
+				terminate = true
+			} else {
+				for _, buddy := range parent.GetChildren() {
+					if buddy != c {
```

Comment:
>if buddy != c { [](start = 5, length = 15)

Compare name?

Reply:
Maybe you can give a Equal method of cell, so you will not forget.

---
In reply to: [335963626](https://github.com/microsoft/pai/pull/3684#discussion_r335963626) [](ancestors = 335963626)

Reply:
Seems cellA == CellB cannot be replaced by cellA.Name == CellB.Name. For example, CellB may be null. How will you handle this? Make a CellEqual(c1, c2)?

---
In reply to: [335964282](https://github.com/microsoft/pai/pull/3684#discussion_r335964282) [](ancestors = 335964282,335963626)